Fix mqtt light brightness slider

thinkl33t opened the pull request:
## Description:

RGB scaling should not be calculated when you set an mqtt brightness topic (IE you want to send a 'set the brightness to 50%' command directly), in which case a 'go red' and a 'set brightness to 50%' topic should both be sent.

If you don't set a brightness topic, the brightness slider should still appear in homeassistant, but the RGB values will be scaled and sent through to the RGB topic.

This has not been happening, the brightness slider wasn't showing, and RGB updates were only storing the colour, not the brightness.  This PR fixes the issue.

**Related issue (if applicable):** fixes #15964

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

thinkl33t committed:
Enable brightness slider for RGB

If we are using RGB with no brightness topic, the brighness slider
should still be visible, as we can scale the RGB amount to give us the
brightness.

thinkl33t committed:
Output RGB scaled by brightness

If we are outputting to an RGB device, but do not have a dedicated
brightness topic set, when the brightness slider is changed, we should
output the current colour's HS, with the V coming from the brightness
slider.

thinkl33t committed:
Brightness from RGB when we're not using a brightness topic

When we aren't using a brightness topic, set the brightness slider based
on the received value from an RGB -> HSV conversion.

thinkl33t committed:
Test for new brightness state scaled by RGB

This adds a test to make sure the brightness stored in the state is
being computed correctly from the RGB value when a dedicated brightness
topic is not set.

armills commented on homeassistant/components/light/mqtt.py:
```diff
@@ -223,6 +223,8 @@ def __init__(self, name, unique_id, effect_list, topic, templates,
         self._supported_features |= (
             topic[CONF_RGB_COMMAND_TOPIC] is not None and SUPPORT_COLOR)
+        self._supported_features |= (
+            topic[CONF_RGB_COMMAND_TOPIC] is not None and SUPPORT_BRIGHTNESS)
```
**Preference:**
Let's just roll this into the line above: `and (SUPPORT_COLOR | SUPPORT_BRIGHTNESS)`.

armills commented on homeassistant/components/light/mqtt.py:
```diff
@@ -563,6 +569,27 @@ def supported_features(self):
             if self._optimistic_brightness:
                 self._brightness = kwargs[ATTR_BRIGHTNESS]
                 should_update = True
+        elif ATTR_BRIGHTNESS in kwargs and \
```
We should include `ATTR_HS_COLOR not in kwargs` here, since if both color and brightness are specified in a request, the RGB topic will have already been published with this brightness taken into account, so we don't want to publish twice.

armills commented on homeassistant/components/light/mqtt.py:
```diff
@@ -563,6 +569,27 @@ def supported_features(self):
+                self.hass, self._topic[CONF_RGB_COMMAND_TOPIC],
+                rgb_color_str, self._qos, self._retain)
+
+            if self._optimistic_brightness:
```
We also need to make sure `self._optimistic_brightness` is set correctly when operating like this. Currently it's based on the existence of [the brightness state topic](https://github.com/home-assistant/home-assistant/blob/06bb69003745db191a514db39fb4afd20c9d25de/homeassistant/components/light/mqtt.py#L205), but when we're using RGB for brightness control we need to also check if the RGB state topic is defined.

thinkl33t replied:
I've gotten a bit tied up trying to implement this - is it actually the RGB state topic we need to be checking, or the existence of the RGB command topic / Non-existence of the brightness command topic?

I'm currently thinking we want optimistic brightness if:

```
we are running in optimistic mode
OR
the brightness state topic isn't set and we have a brightness control topic
OR
the rgb state topic isn't set and we don't have a brightness control topic
```

Does this sound correct?

thinkl33t replied:
I believe this is done as of https://github.com/home-assistant/home-assistant/pull/17075/commits/faabbd4ab701cd7ab33e5e4b21dc26d758a29ca8

emlove reviewed (changes requested):
The feature looks good! I've added some notes with some implementation concerns.

We'll also need to add tests for the `turn_on` service, to test that the RGB topic is published for this mode.

thinkl33t committed:
Changes from review

Fixes formatting of supported features flags, and checks HS colour
hasn't been set when operating in RGB-only mode

homeassistant commented:
Hello @thinkl33t,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`e74dfb4966da0b27c1424b528ec4a4be468c8095`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;17075&#x2F;commits&#x2F;e74dfb4966da0b27c1424b528ec4a4be468c8095) - This commit has something that looks like an email address (thinkl33t@bc-laptop.office.codethink.co.uk). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

thinkl33t committed:
Set optimistic brightness correctly in rgb mode

When we're using rgb mode to set the brightness, we want to set
optimistic brightness if:

we are running in optimistic mode
OR
the brightness state topic isn't set and we have a brightness command topic
OR
the rgb state topic isn't set and we don't have a brightness command topic

thinkl33t committed:
Add test for turn_on in RGB brightness mode

thinkl33t commented:
I believe this is now ready for review / merge :)

emlove reviewed (approved):
This LGTM! Nice work on the tests as well!
